### DISCUSSION TITLE
Thoughts on collection of boilerplates for somewhat common variables

### DISCUSSION THREAD

**Discussion by bradenhilton** [twpayne/chezmoi#1439] [General]
Recently I noticed the section on determining the chassis type of the current machine was added to the how to, and it got me thinking about similar things that might be nice to have as variables.

I was wondering what everyone thought about perhaps creating some kind of centralized repo containing boilerplates/snippets that can be used to achieve similar things. I'm mainly focusing on things that can be determined in an abstract way (if the variable itself was highly specific to your needs and your machine(s), there would be little point in adding it to a repo like this) and can also be done on all or most platforms supported by chezmoi.

Some variables that would be nice to have for me personally that I can think of off the top of my head:

- Whether the current machine is a laptop/desktop/mobile - already achieved with the chassis type snippet plus an additional check `if eq .chezmoi.os "android"`.
- The CPU core count of the current machine - mpv assigns duplicate copies of the thumbnailer script that I use to additional threads in order to generate thumbnails faster. This can bog down an older machine if the number of thumbnailer threads is greater than ~1/2 of the core count and the CPU is also being used to play the video.
- The number of displays connected to/active on the current machine - I typically use three displays with my desktop machine (but not always), and no external displays with my ThinkPad or MacBook Pro. Many applications allow settings to the effect of "Always open on display X" etc.

It also occurred to me that some of these could be implemented in chezmoi itself, depending on demand/ease of implementation/scope of course.

Feel free to offer your thoughts and also any ideas for useful variables you might have.

**Comment by twpayne** [comment]
This is a great idea!

As you say, there's an initial start in the [useful templates doc](https://github.com/twpayne/chezmoi/blob/master/docs/TEMPLATING.md#useful-templates), but I think it's worth separating this into its own documentation.

A few practices that I think are worth including:
* Using boolean variable tags to control your dotfiles at a high level.
* Working around macOS's unreliable hostname.
* Detecting ephemeral environments, e.g. docker containers, vagrant boxes, and GitHub codespaces.

**Reply by bradenhilton** [reply]
Do you have any ideas regarding...

* How to structure the templates (separate files in a templates directory, single file with templates separated by functionality etc.)
* How to bring attention to this change, to encourage and emphasize that users are welcome to share their templates if they think they might be useful
* If templates that are submitted should focus solely on performance over readability, or vice-versa, or perhaps both and let users pick whichever one they care about most
* If there should be any kind of minimum requirement where appropriate, e.g. templates which achieve something that can be done on all platforms must provide complete implementations for all platforms when submitted, or if a single platform implementation with TODOs are acceptable
* If templates that are submitted should have some form of testing, to ensure that future changes to their implementation (if any) still provide the intended values (I have absolutely no idea off the top of my head how this could be achieved)

Also, would it be worth opening an issue (perhaps an RFC) regarding this to get more eyes on it?

**Comment by twpayne** [comment]
Let's convert this to an issue for more visibility.
